storages/redis/redis.go

```diff
@@ -76,71 +84,89 @@ func (s *Storage) ResetState(chatId, userId int64, withData bool) error {
+
+	for {
+		var err error
+		keys, cursor, err = s.rds.Scan(
+			context.TODO(),
+			cursor,
+			s.generateKey(chatId, userId, stateDataKey, "*"),
+			s.resetDataBatchSize,
+		).Result()
+		if err != nil {
+			return err
+		}
+
+		if len(keys) > 0 {
+			if err := s.rds.Del(context.TODO(), keys...).Err(); err != nil {
+				return err
+			}
 		}
-	}
 
-	// unmarshal state data
-	var stateData map[string]interface{}
-	if err := json.Unmarshal(stateDataBytes, &stateData); err != nil {
-		return fmt.Errorf("unmarshal state data: %w", err)
+		if cursor == 0 {
+			break
+		}
 	}
 
-	// set or delete data
+	return nil
+}
+
+func (s *Storage) UpdateData(chatId, userId int64, key string, data interface{}) error {
 	if data == nil {
-		delete(stateData, key)
-	} else {
-		stateData[key] = data
+		return s.rds.Del(context.TODO(), s.generateKey(chatId, userId, stateDataKey, key)).Err()
 	}
 
-	// marshal state data
-	stateDataBytes, err = json.Marshal(stateData)
+	encodedData, err := s.encode(data)
 	if err != nil {
-		return fmt.Errorf("marshal state data: %w", err)
+		return err
 	}
-
-	return s.rds.Set(context.TODO(), s.generateKey(chatId, userId, stateDataKey), stateDataBytes, s.ttlData).Err()
+	return s.rds.Set(context.TODO(), s.generateKey(chatId, userId, stateDataKey, key), encodedData, s.ttlData).Err()
 }
 
-func (s *Storage) GetData(chatId, userId int64, key string) (interface{}, error) {
-	stateDataBytes, err := s.rds.Get(context.TODO(), s.generateKey(chatId, userId, stateDataKey)).Bytes()
+func (s *Storage) GetData(chatId, userId int64, key string, to interface{}) error {
+	dataBytes, err := s.rds.Get(context.TODO(), s.generateKey(chatId, userId, stateDataKey, key)).Bytes()
 	if errors.Is(err, redis.Nil) {
-		return nil, fsm.ErrNotFound
+		return fsm.ErrNotFound
 	}
 	if err != nil {
-		return nil, fmt.Errorf("get state data: %w", err)
-	}
-
-	// unmarshal state data
-	var stateData map[string]interface{}
-	if err := json.Unmarshal(stateDataBytes, &stateData); err != nil {
-		return nil, fmt.Errorf("unmarshal state data: %w", err)
+		return fmt.Errorf("get data: %w", err)
 	}
 
-	// get data
-	res, ok := stateData[key]
-	if !ok {
-		return nil, fsm.ErrNotFound
-	}
-
-	return res, nil
+	return s.decode(dataBytes, to)
 }
 
 func (s *Storage) Close() error {
 	return s.rds.Close()
 }
 
-func (s *Storage) generateKey(chat, user int64, keyType keyType) string {
-	return fmt.Sprintf("%s:%d:%d:%s", s.prefix, chat, user, keyType)
+func (s *Storage) generateKey(chat, user int64, keyType keyType, keys ...string) string {
+	res := fmt.Sprintf("%s:%d:%d:%s", s.prefix, chat, user, keyType)
+	if keys != nil {
+		res += ":" + strings.Join(keys, ":")
+	}
+	return res
+}
+
+func (s *Storage) encode(data interface{}) ([]byte, error) {
+	var buf bytes.Buffer
+	encoder := gob.NewEncoder(&buf)
+
+	if err := encoder.Encode(data); err != nil {
+		return nil, err
+	}
+
+	return buf.Bytes(), nil
+}
+
+func (s *Storage) decode(data []byte, to interface{}) error {
```

Comment:
why gob if json also decodes well into structures? I saw your message in #3 about numbers in json, but it doesn't matter in this case

Reply:
> why gob if json also decodes well into structures? I saw your message in #3 about numbers in json, but it doesn't matter in this case

In #3 I did saving the data like in aiogram:
```
"fsm:1:1:data": "{'age': 21, 'name': 'Peter'}"
```

Here I did the saving data another way:
```
"fsm:1:1:data:age": "21"
"fsm:1:1:data:name": "Peter"
```

In the first variant I think it'll be hard to save the data in `to interface{}` argument in `GetData` method, because it'll need a lot of reflection.

In the second variant we can decode/encode the data in any way we want actually. I've just picked gob because it's faster and simpler to use.
I see the only one reason to decode/encode the data via JSON - make the values readable in redis-cli :)